### PR TITLE
Add various REST API docs (Reports, Text Messages, Contacts)

### LINKED ISSUE #68: Write documentation on the Reports component
We need to bring over the documentation for the reports section and update it, converting to RST format.

https://developer.mautic.org/#reports

https://mautic-developer.readthedocs.io/en/latest/components/reports.html

### LINKED ISSUE #70: Write documentation for Text Messaging component
We need to bring over, update and convert to RST the documentation on sending text messages.

https://developer.mautic.org/#text-messages

https://mautic-developer.readthedocs.io/en/latest/components/sms.html

### PR DESCRIPTION
**Pull request by dennisameling**
Closes #70 
Closes #68 

@RCheesley FYI, the old Developer Docs had a section on REST API which the new one doesn't have yet. The link in #70 referred to the REST API section, so I added that section in this PR.

Ideally Mautic should auto-generate these docs through OpenAPI, but we're not there yet. This is the best we can do for now.

I used https://github.com/miyakogi/m2r to convert the existing Markdown page over to RST. It did a lot of stuff for me already (including `list-table`, YAY!). I just had to update headings and address a bunch of Vale errors/warnings/suggestions, but this was quite easy to do.

<a href="https://gitpod.io/#https://github.com/mautic/developer-documentation-new/pull/90"></a>

### CODE REVIEW

**Comment by dennisameling** [comment]
Just added the REST API docs for Contacts as well. That's by far the largest one, so the other ones should be quite a bit faster to migrate. Sorry for the large PR 😅

File: docs/rest_api/contacts.rst [hunk 1]
```diff
@@ -0,0 +1,2300 @@
...
+.. code-block:: php
+
+   <?php
+   use Mautic\MauticApi;
+   use Mautic\Auth\ApiAuth;
+
+   // ...
+   $initAuth   = new ApiAuth();
+   $auth       = $initAuth->newAuth($settings);
+   $apiUrl     = "https://your-mautic.com";
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:19]:
```suggestion
   $apiUrl     = "https://example.com";
```

File: docs/rest_api/contacts.rst [hunk 2]
```diff
@@ -0,0 +1,2300 @@
...
+       'val' => '444444444,888888888',
+     ]
+   ];
+
+This design allows to add multiple conditions in the same request.
+
+If you aren't using PHP, here is URL-encoded version of the example:
+``GET https://[your_mauitc_domain]/api/contacts?where%5B0%5D%5Bcol%5D=phone&where%5B0%5D%5Bexpr%5D=in&where%5B0%5D%5Bval%5D=444444444,888888888``
+
+`List of available expressions <https://www.doctrine-project.org/projects/doctrine-orm/en/2.7/reference/query-builder.html#the-expr-class>`_
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:273]: Should this be an external link?

File: docs/rest_api/contacts.rst [hunk 3]
```diff
@@ -0,0 +1,2300 @@
...
+       'col' => 'phone',
+       'expr' => 'in',
+       'val' => '444444444,888888888',
+     ]
+   ];
+
+This design allows to add multiple conditions in the same request.
+
+If you aren't using PHP, here is URL-encoded version of the example:
+``GET https://[your_mauitc_domain]/api/contacts?where%5B0%5D%5Bcol%5D=phone&where%5B0%5D%5Bexpr%5D=in&where%5B0%5D%5Bval%5D=444444444,888888888``
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:271]:
```suggestion
``GET https://[example.com]/api/contacts?where%5B0%5D%5Bcol%5D=phone&where%5B0%5D%5Bexpr%5D=in&where%5B0%5D%5Bval%5D=444444444,888888888``
```

File: docs/rest_api/contacts.rst [hunk 4]
```diff
@@ -0,0 +1,2300 @@
...
+.. vale on
+
+.. code-block:: php
+
+   <?php
+
+   $data = array(
+       'firstname' => 'Jim',
+       'lastname'  => 'Contact',
+       'email'     => 'jim@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:415]:
```suggestion
       'email'     => 'jim@example.com',
```

File: docs/rest_api/contacts.rst [hunk 5]
```diff
@@ -0,0 +1,2300 @@
...
+
+.. code-block:: php
+
+   <?php
+
+   $data = array(
+       array(
+       'firstname' => 'Jim',
+       'lastname'  => 'Contact',
+       'email'     => 'jim@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:474]:
```suggestion
       'email'     => 'jim@example.com',
```

File: docs/rest_api/contacts.rst [hunk 6]
```diff
@@ -0,0 +1,2300 @@
...
+       'ipAddress' => $_SERVER['REMOTE_ADDR']
+       ),
+       array(
+           'firstname' => 'John',
+       'lastname'  => 'Doe',
+       'email'     => 'john@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:480]:
```suggestion
       'email'     => 'rudolf@example.com',
```

File: docs/rest_api/contacts.rst [hunk 7]
```diff
@@ -0,0 +1,2300 @@
...
+
+.. vale on
+
+.. code-block:: php
+
+   <?php
+
+   $id   = 1;
+   $data = array(
+       'email'     => 'jim-new-address@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:534]:
```suggestion
       'email'     => 'isabel-new-address@example.com',
```

File: docs/rest_api/contacts.rst [hunk 8]
```diff
@@ -0,0 +1,2300 @@
...
+.. code-block:: php
+
+   <?php
+
+   $data = [
+       [
+           'id'        => 1,
+           'firstname' => 'Jim',
+           'lastname'  => 'Contact',
+           'email'     => 'jim@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:607]:
```suggestion
           'email'     => 'jim@example.com',
```

File: docs/rest_api/contacts.rst [hunk 9]
```diff
@@ -0,0 +1,2300 @@
...
+           'ipAddress' => $_SERVER['REMOTE_ADDR']
+       ],
+       [
+           'id'        => 1,
+           'firstname' => 'John',
+           'lastname'  => 'Doe',
+           'email'     => 'john@his-site.com',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:614]:
```suggestion
           'email'     => 'ashish@example.com',
```

File: docs/rest_api/contacts.rst [hunk 10]
```diff
@@ -0,0 +1,2300 @@
...
+
+**HTTP Request**
+
+.. vale on
+
+``DELETE /contacts/batch/delete``
+
+If you aren't using PHP, here is a URL example:
+
+``DELETE https://[your_mautic_domain]/api/contacts/batch/delete?ids=1,2``
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:725]:
```suggestion
``DELETE https://[example.com]/api/contacts/batch/delete?ids=1,2``
```

File: docs/rest_api/contacts.rst [hunk 11]
```diff
@@ -0,0 +1,2300 @@
...
+             "campaignName": "Campaign Date"
+           }
+         }
+       },
+       {
+         "event": "email.sent",
+         "eventId": "email.sent796",
+         "eventLabel": {
+           "label": "2017-05-23 - Email - Leads - Nurture Flow (Monica) 1",
+           "href": "http:\/\/mautic.dev\/email\/view\/597a116ae69ca",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1865]:
```suggestion
           "href": "http:\/\/example.com\/email\/view\/597a116ae69ca",
```

File: docs/rest_api/contacts.rst [hunk 12]
```diff
@@ -0,0 +1,2300 @@
...
+           },
+           "type": "sent"
+         }
+       },
+       {
+         "event": "email.read",
+         "eventId": "email.read769",
+         "eventLabel": {
+           "label": "Custom Email: device test",
+           "href": "http:\/\/mautic.dev\/email\/view\/5966b0cd571f4",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1895]:
```suggestion
           "href": "http:\/\/example.com\/email\/view\/5966b0cd571f4",
```

File: docs/rest_api/contacts.rst [hunk 13]
```diff
@@ -0,0 +1,2300 @@
...
+             "ipAddress": {
+               "ip": "127.0.0.1"
+             },
+             "form": {
+               "id": 143,
+               "name": "3586 Test",
+               "alias": "3586_test"
+             },
+             "dateSubmitted": "2017-07-27T03:09:07+00:00",
+             "referer": "http:\/\/mautic.dev\/form\/143",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1960]:
```suggestion
             "referer": "http:\/\/example.com\/form\/143",
```

File: docs/rest_api/contacts.rst [hunk 14]
```diff
@@ -0,0 +1,2300 @@
...
+             "results": {
+               "form_id": "143",
+               "email": "formtest7@test.com",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1963]:
```suggestion
               "email": "formtest7@example.com",
```

File: docs/rest_api/contacts.rst [hunk 15]
```diff
@@ -0,0 +1,2300 @@
...
+           "href": "\/s\/pages\/view\/8"
+         },
+         "eventType": "Page hit",
+         "timestamp": "2017-07-21T20:36:49+00:00",
+         "contactId": "16380",
+         "details": {
+           "hit": {
+             "UserAgent": "Mozilla\/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/59.0.3071.115 Safari\/537.36",
+             "dateHit": "2017-07-21T20:36:49+00:00",
+             "url": "http:\/\/mautic.dev\/uncategorized\/translation-test1",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1988]:
```suggestion
             "url": "http:\/\/example.com\/uncategorized\/translation-test1",
```

File: docs/rest_api/contacts.rst [hunk 16]
```diff
@@ -0,0 +1,2300 @@
...
+             "query": {
+               "pageUrl": "http:\/\/mautic.dev\/uncategorized\/translation-test1"
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:1990]:
```suggestion
               "pageUrl": "http:\/\/example.com\/uncategorized\/translation-test1"
```

File: docs/rest_api/contacts.rst [hunk 17]
```diff
@@ -0,0 +1,2300 @@
...
+         "eventType": "Asset downloaded",
+         "timestamp": "2017-04-04T01:49:13+00:00",
+         "details": {
+           "asset": {
+             "id": 1,
+             "title": "Download Mautic",
+             "alias": "download-mautic",
+             "description": "test"
+           },
+           "assetDownloadUrl": "http:\/\/mautic.dev\/asset\/1:download-mautic"
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:2059]:
```suggestion
           "assetDownloadUrl": "http:\/\/example.com\/asset\/1:download-mautic"
```

File: docs/rest_api/contacts.rst [hunk 18]
```diff
@@ -0,0 +1,2300 @@
...
+
+   {
+     "total":1,
+     "companies":[
+       {
+         "company_id":"420",
+         "date_associated":"2016-12-27 15:03:43",
+         "is_primary":"0",
+         "companyname":"test",
+         "companyemail":"test@company.com",
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:2190]:
```suggestion
         "companyemail":"test@example.com",
```

File: docs/rest_api/reports.rst [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+.. code-block:: php
+
+   <?php
+   use Mautic\MauticApi;
+   use Mautic\Auth\ApiAuth;
+
+   // ...
+   $initAuth  = new ApiAuth();
+   $auth      = $initAuth->newAuth($settings);
+   $apiUrl    = "https://your-mautic.com";
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:19]:
```suggestion
   $apiUrl    = "https://example.com";
```

File: docs/rest_api/reports.rst [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+``Expected Response Code: 200``
+
+.. code-block:: json
+
+   {
+     "totalResults": 3990,
+     "data": [
+       {
+         "id2": "12",
+         "email1": "john@doe.email",
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:70]:
```suggestion
         "email1": "john@example.com",
```

File: docs/rest_api/reports.rst [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+         "firstname1": "",
+         "lastname1": ""
+       },
+       {
+         "id2": "23",
+         "email1": "johnatan@doe.email",
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:76]:
```suggestion
         "email1": "alex@example.com",
```

File: docs/rest_api/reports.rst [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+         "firstname1": "",
+         "lastname1": ""
+       },
+       {
+         "id2": "24",
+         "email1": "johny@doe.email",
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:82]:
```suggestion
         "email1": "amal@example.com",
```

File: docs/rest_api/reports.rst [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+         "firstname1": "",
+         "lastname1": ""
+       },
+       {
+         "id2": "25",
+         "email1": "johan@doe.email",
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:88]:
```suggestion
         "email1": "ariel@example.com",
```

File: docs/rest_api/reports.rst [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+         "firstname1": "",
+         "lastname1": ""
+       },
+       {
+         "id2": "26",
+         "email1": "jane@doe.email",
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:94]:
```suggestion
         "email1": "bola@example.com",
```

File: docs/rest_api/reports.rst (hunk 20 above)
Comment (RCheesley) [reply docs/rest_api/reports.rst:70]: Let's be more inclusive with the name examples we use. I selected some from https://developers.google.com/style/examples#example-person-names

File: docs/rest_api/contacts.rst [hunk 25]
```diff
@@ -0,0 +1,2300 @@
...
+
+   $data = array(
+       array(
+       'firstname' => 'Jim',
+       'lastname'  => 'Contact',
+       'email'     => 'jim@his-site.com',
+       'ipAddress' => $_SERVER['REMOTE_ADDR']
+       ),
+       array(
+           'firstname' => 'John',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:478]:
```suggestion
           'firstname' => 'Rudolf',
```

File: docs/rest_api/contacts.rst [hunk 26]
```diff
@@ -0,0 +1,2300 @@
...
+       'lastname'  => 'Doe',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:479]:
```suggestion
       'lastname'  => 'Große',
```

File: docs/rest_api/contacts.rst [hunk 27]
```diff
@@ -0,0 +1,2300 @@
...
+       [
+           'id'        => 1,
+           'firstname' => 'Jim',
+           'lastname'  => 'Contact',
+           'email'     => 'jim@his-site.com',
+           'ipAddress' => $_SERVER['REMOTE_ADDR']
+       ],
+       [
+           'id'        => 1,
+           'firstname' => 'John',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:612]:
```suggestion
           'firstname' => 'Ashish',
```

File: docs/rest_api/contacts.rst [hunk 28]
```diff
@@ -0,0 +1,2300 @@
...
+           'lastname'  => 'Doe',
```
Comment (RCheesley) [inline docs/rest_api/contacts.rst:613]:
```suggestion
           'lastname'  => 'Wallach',
```

File: docs/rest_api/reports.rst [hunk 29]
```diff
@@ -0,0 +1,311 @@
...
+**********
+
+.. vale on
+
+.. code-block:: php
+
+   <?php
+
+   //...
+   // Simple get all with default options:
```
Comment (RCheesley) [inline docs/rest_api/reports.rst:35]:
```suggestion
   // Get all with default options:
```

File: docs/rest_api/text_messages.rst [hunk 30]
```diff
@@ -0,0 +1,372 @@
+Text messages
+#############
+
+Use this endpoint to obtain details on Mautic's Text Messages, or SMSes.
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:4]:
```suggestion
Use this endpoint to obtain details on Mautic's Text Messages, or SMS's.
```

Comment (RCheesley) [reply docs/rest_api/text_messages.rst:4]: I think it should be apostrophe'd but I am not 100% sure - open to discussion!

File: docs/rest_api/text_messages.rst [hunk 31]
```diff
@@ -0,0 +1,372 @@
...
+.. code-block:: php
+
+   <?php
+   use Mautic\MauticApi;
+   use Mautic\Auth\ApiAuth;
+
+   // ...
+   $initAuth = new ApiAuth();
+   $auth     = $initAuth->newAuth($settings);
+   $apiUrl   = "https://your-mautic.com";
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:73]:
```suggestion
   $apiUrl   = "https://example.com";
```

File: docs/rest_api/text_messages.rst [hunk 32]
```diff
@@ -0,0 +1,372 @@
...
+``Expected Response Code: 200``
+
+.. code-block:: json
+
+   {  
+       "sms":{  
+           "isPublished":true,
+           "dateAdded":"2016-09-14T12:14:45+00:00",
+           "createdBy":1,
+           "createdByUser":"John Doe",
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:108]:
```suggestion
           "createdByUser":"Wu Popovski",
```

File: docs/rest_api/text_messages.rst [hunk 33]
```diff
@@ -0,0 +1,372 @@
...
+.. code-block:: json
+
+   {  
+       "total":1,
+       "smses":[  
+           {  
+               "isPublished":true,
+               "dateAdded":"2016-09-14T12:14:45+00:00",
+               "createdBy":1,
+               "createdByUser":"John Doe",
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:176]:
```suggestion
               "createdByUser":"Kevin Bulgarelli",
```

File: docs/rest_api/text_messages.rst [hunk 34]
```diff
@@ -0,0 +1,372 @@
...
+
+   <?php
+
+   $id   = 1;
+   $data = array(
+       'name'        => 'New sms name',
+       'isPublished' => 0
+   );
+
+   // Create new a sms of ID 1 is not found?
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:246]:
```suggestion
   // Create new a SMS of ID 1 is not found?
```

File: docs/rest_api/text_messages.rst [hunk 35]
```diff
@@ -0,0 +1,372 @@
...
+**PUT** creates an SMS if the given ID doesn't exist and clears all the SMS information, adds the information from the request.
+**PATCH** fails if the SMS with the given ID doesn't exist and updates the SMS field values with the values from the request.
+
+.. vale off
+
+**HTTP Request**
+
+.. vale on
+
+To edit a SMS and return a 404 if the SMS isn't found:
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:262]:
```suggestion
To edit an SMS and return a 404 if the SMS isn't found:
```

File: docs/rest_api/text_messages.rst [hunk 36]
```diff
@@ -0,0 +1,372 @@
...
+
+``PATCH /smses/ID/edit``
+
+To edit a SMS and create a new one if the SMS isn't found:
```
Comment (RCheesley) [inline docs/rest_api/text_messages.rst:266]:
```suggestion
To edit an SMS and create a new one if the SMS isn't found:
```

File: docs/rest_api/text_messages.rst (hunk 30 above)
Comment (dennisameling) [reply docs/rest_api/text_messages.rst:4]: There's lots of discussion about this [here](https://english.stackexchange.com/questions/51378/what-is-the-plural-form-of-sms), but the general consensus seems to be on `SMSs` or `texts`. WDYT?

Comment (RCheesley) [reply docs/rest_api/text_messages.rst:4]: I'd go with SMSs as we refer to Text Messages immediately prior.

Comment (dennisameling) [reply docs/rest_api/text_messages.rst]: Updated. Unfortunately the endpoint itself is `GET /smses/ID` so we can't update that part of the docs. It'd require a code update in `mautic/mautic` first.

Comment (RCheesley) [reply docs/rest_api/text_messages.rst]: Yeah that will have to stay as it is for now! Thanks!

**Comment by dennisameling** [comment]
@RCheesley this is good to merge now, right?

**Pull request merged**